### PR TITLE
feat(deferral): created deferral journey

### PR DESCRIPTION
**Pull request by alittlecross**
- added locales files for page content
- updated /changes-enquiries/personal page
- created unit tests for /stop-state-pension functions
- created route and functions for /stop-state-pension
- updated fromValidation functions
- created page for /stop-state-pension
- updated /death page and functions
- created unit tests for /deferral functions
- created route and functions for /deferral
- updated fromValidation functions
- created pages for /deferral
- updated failing tests

Jira SA 2966

Signed-off-by: Paul Fazackerley <paul.fazackerley@engineering.digital.dwp.gov.uk>

### CODE REVIEW

File: app/routes/changes-enquiries/deferral/functions.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+  const recordDeferralApiUri = 'api/award/record-deferral';
+  const putCall = requestHelper.generatePutCall(res.locals.agentGateway + recordDeferralApiUri, deferralDetails, 'award', req.user);
+  request(putCall)
+    .then(() => {
+      req.flash('success', i18n.t('deferral-update:messages.success'));
+      deleteSession.deleteAllDeferralSession(req);
+      res.redirect(root);
+    })
+    .catch((err) => {
+      const traceID = requestHelper.getTraceID(err);
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/deferral/functions.js:112]: Have a little helper that does all this for you:

```
errorHelper.flashErrorAndRedirect(req, res, err, 'award', `${root}/deferral/update`);
```

File: app/routes/changes-enquiries/deferral/functions.js [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+function getUpdate(req, res) {
+  const { nino } = dataStore.get(req, 'awardDetails');
+  const fromDate = dataStore.get(req, 'from-date', 'deferral');
+  const deferralDetails = deferralObject.formatter(nino, fromDate);
+  const recordDeferralApiUri = 'api/award/record-deferral';
+  const putCall = requestHelper.generatePutCall(res.locals.agentGateway + recordDeferralApiUri, deferralDetails, 'award', req.user);
+  request(putCall)
+    .then(() => {
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/deferral/functions.js:106]: Can we use the try/catch await approach for these, instead of using `then`

File: app/routes/changes-enquiries/death/functions.js [hunk 3]
```diff
@@ -34,9 +34,10 @@ const [CANNOT_CALCULATE, OVERPAYMENT, ARREARS, NOTHING_OWED, DEATH_NOT_VERIFIED]
 
 function getAddDateDeath(req, res) {
   const { awardDetails } = req.session;
+  const keyDetails = keyDetailsHelper.formatter(awardDetails);
   const details = dataStore.get(req, 'date-of-death', 'death');
   res.render('pages/changes-enquiries/death/enter-date', {
-    awardDetails,
+    keyDetails,
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/death/functions.js:40]: Remember you talking about this but wasn't quite sure what you meant. But we don't actually need to do this as keyDetails is passed to the locals and to the views via middleware. If this exists in other functions you can remove it. The middleware was an afterthought so went through all the code to remove `keyDetails` but might have missed a few

File: app/routes/changes-enquiries/death/functions.js [hunk 4]
```diff
@@ -34,9 +34,10 @@ const [CANNOT_CALCULATE, OVERPAYMENT, ARREARS, NOTHING_OWED, DEATH_NOT_VERIFIED]
 
 function getAddDateDeath(req, res) {
   const { awardDetails } = req.session;
+  const keyDetails = keyDetailsHelper.formatter(awardDetails);
   const details = dataStore.get(req, 'date-of-death', 'death');
   res.render('pages/changes-enquiries/death/enter-date', {
-    awardDetails,
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/death/functions.js:39]: Why was this removed

File: app/routes/changes-enquiries/award/functions.js [hunk 5]
```diff
@@ -12,7 +12,7 @@ const secondaryNavigationList = secondaryNavigationHelper.navigationItems(active
 
 async function getAwardList(req, res) {
   try {
-    const award = await dataStore.cacheRetriveAndStore(req, null, 'awardDetails', () => {
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/award/functions.js:15]: Ahh...a good example of my ability to spell

File: app/routes/changes-enquiries/deferral/functions.js [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+const root = '/changes-and-enquiries/personal';
+
+const dateRequestReceivedUrl = `${root}/deferral/date-request-received`;
+const defaultDateUrl = `${root}/deferral/default-date`;
+const stopStatePensionUrl = `${root}/stop-state-pension`;
+
+function getDateRequestReceived(req, res) {
+  const awardDetails = dataStore.get(req, 'awardDetails');
+  const details = dataStore.get(req, 'date-request-received', 'deferral');
+  const keyDetails = keyDetailsHelper.formatter(awardDetails);
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/deferral/functions.js:22]: Remove keyDetails, same reason as the comment above and for all the other functions

File: app/routes/changes-enquiries/deferral/functions.js [hunk 7]
```diff
@@ -0,0 +1,124 @@
+const i18n = require('i18next');
+const request = require('request-promise');
+
+const dataStore = require('../../../../lib/dataStore');
+const dateHelper = require('../../../../lib/dateHelper');
+const deferralObject = require('../../../../lib/objects/deferralObject');
+const deleteSession = require('../../../../lib/deleteSession');
+const formValidator = require('../../../../lib/formValidator');
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/deferral/functions.js:8]: Could you split out all the deferral validation into its own file, I've started doing this over this over the last year as that `formValidator` file is getting far too big

File: app/routes/changes-enquiries/deferral/functions.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+      keyDetails,
+    });
+  }
+}
+
+function getDefaultDate(req, res) {
+  const awardDetails = dataStore.get(req, 'awardDetails');
+  const defaultDate = dataStore.get(req, 'default-date', 'deferral');
+  const keyDetails = keyDetailsHelper.formatter(awardDetails);
+  const { statePensionDate } = dataStore.get(req, 'awardDetails');
```
Comment (malcolmhire) [inline app/routes/changes-enquiries/deferral/functions.js:54]: You've already got `awardDetails` session above. But when you remove `keyDetails` this can stay as is

File: app/views/pages/changes-enquiries/deferral/default-date.html [hunk 9]
```diff
@@ -0,0 +1,47 @@
+{% extends 'pages/changes-enquiries/layout.html' %}
+
+{% from "govuk/components/back-link/macro.njk" import govukBackLink %}
+{% from "govuk/components/inset-text/macro.njk" import govukInsetText %}
+{% from "govuk/components/radios/macro.njk" import govukRadios %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+
+{% block pageTitle %}{{ t("deferral-default-date:header") }} - Changes and enquiries{% endblock %}
```
Comment (malcolmhire) [inline app/views/pages/changes-enquiries/deferral/default-date.html:8]:
```suggestion
{% block pageTitle %}{{ t("deferral-default-date:header") }} - {{ t('app:title.changes-enquiries') }}{% endblock %}
```

File: app/views/pages/changes-enquiries/stop-state-pension/index.html [hunk 10]
```diff
@@ -0,0 +1,41 @@
+{% extends 'pages/changes-enquiries/layout.html' %}
+
+{% from "govuk/components/back-link/macro.njk" import govukBackLink %}
+{% from "govuk/components/radios/macro.njk" import govukRadios %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+
+{% block pageTitle %}{{ t("stop-state-pension:header") }} - Changes and enquiries{% endblock %}
```
Comment (malcolmhire) [inline app/views/pages/changes-enquiries/stop-state-pension/index.html:7]:
```suggestion
{% block pageTitle %}{{ t("stop-state-pension:header") }} - {{ t('app:title.changes-enquiries') }}{% endblock %}
```

File: app/views/pages/customer/add.html [hunk 11]
```diff
@@ -70,7 +70,7 @@ <h1 class="govuk-heading-xl no-js">Add Customer</h1>
             type: "text",
             classes: "govuk-!-width-one-half",
             label: {
-              text: t("add:field_names.firstname"),
+              text: t("add:field_names.firstName"),
```
Comment (malcolmhire) [inline app/views/pages/customer/add.html:73]: You'll have broken this label here

File: lib/customerFieldsValidator.js [hunk 12]
```diff
@@ -123,13 +123,13 @@ module.exports = {
   isValidFirstName(firstName) {
     let error;
     if (firstName === '') {
-      error = i18n.t('add:errors.firstname.required');
```
Comment (malcolmhire) [inline lib/customerFieldsValidator.js:126]: Same with all this and all other firstNames below

File: lib/objects/reviewAwardReasonObject.js [hunk 13]
```diff
@@ -73,8 +73,8 @@ function formatReasonForChange(reasonForChange) {
   case 'COEG INVESTIGATION C.O. POSITION - BP CALCULATION ONLY':
     reason = 'COEG investigation contracted out position - basic pension calculation only';
     break;
-  case 'RP BENEFIT CALCULATION MADE - FEDBACK DETAILS NOT RECEIVED':
-    reason = 'Retirement pension benefit calculation made - fedback details not received';
+  case 'RP BENEFIT CALCULATION MADE - FEEDBACK DETAILS NOT RECEIVED':
```
Comment (malcolmhire) [inline lib/objects/reviewAwardReasonObject.js:76]: Revert this, please. Will investigate if this is a spelling mistake. We are translating an API response here

File: app/routes/changes-enquiries/death/functions.js (hunk 4 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/death/functions.js:39]: it didn't appear to be used by the view but I've put it back since it's outside the story

File: app/routes/changes-enquiries/death/functions.js (hunk 3 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/death/functions.js:40]: removed

File: locales/en/add.json [hunk 14]
```diff
@@ -2,7 +2,7 @@
   "field_names":
   {
     "title": "Title",
-    "firstname": "First Name",
+    "firstName": "First Name",
```
Comment (malcolmhire) [inline locales/en/add.json:5]: Revert

File: locales/en/add.json [hunk 15]
```diff
@@ -54,7 +54,7 @@
       "toShort": "Surname must be at least two characters",
       "toLong": "Surname must be seventy characters or less"
     },
-    "firstname": {
+    "firstName": {
```
Comment (malcolmhire) [inline locales/en/add.json:57]: Revert

File: app/routes/changes-enquiries/deferral/functions.js (hunk 7 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/deferral/functions.js:8]: done

File: locales/en/stop-state-pension.json [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+  "header": "Stop State Pension",
+  "h2": "Reason for stopping the State Pension",
+  "fields": {
+    "reason": {
+      "options": {
+        "death": "Death",
+        "deferral": "Deferral"
+      },
+      "errors": {
+        "required": "Please make a selection"
```
Comment (malcolmhire) [inline locales/en/stop-state-pension.json:11]:
```suggestion
        "required": "Select a reason for stopping the State Pension"
```

File: app/routes/changes-enquiries/deferral/functions.js (hunk 6 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/deferral/functions.js:22]: done

File: locales/en/deferral-date-request-received.json [hunk 17]
```diff
@@ -0,0 +1,14 @@
+{
+  "header": "Date request received",
+  "h2": "Date the request was received",
+  "fields": {
+    "date": {
+      "errors": {
+        "required": "Please enter a date",
```
Comment (malcolmhire) [inline locales/en/deferral-date-request-received.json:7]:
```suggestion
        "required": "Enter the date the request was received",
```

File: locales/en/deferral-date-request-received.json [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+        "incomplete": "Date request received must include a day, month, and year",
```
Comment (malcolmhire) [inline locales/en/deferral-date-request-received.json:8]: Don't have validation for this

File: app/routes/changes-enquiries/deferral/functions.js (hunk 8 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/deferral/functions.js:54]: 👍

File: locales/en/deferral-date-request-received.json [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+        "format": "Enter a real date request received, like 12 2 2019",
```
Comment (malcolmhire) [inline locales/en/deferral-date-request-received.json:9]:
```suggestion
        "format": "Enter a real date, like 12 3 2020",
```

File: locales/en/deferral-date-request-received.json [hunk 20]
```diff
@@ -0,0 +1,14 @@
...
+        "future": "Date request received must be today or in the past"
```
Comment (malcolmhire) [inline locales/en/deferral-date-request-received.json:10]:
```suggestion
        "future": "Date the request was received cannot be in the future"
```

File: locales/en/deferral-default-date.json [hunk 21]
```diff
@@ -0,0 +1,12 @@
+{
+  "header": "Defer from State Pension date?",
+  "h2": "Does the claimant want to defer from their State Pension date?",
+  "insetText": "State Pension date: __DATE__",
+  "fields": {
+    "defaultDate": {
+      "errors": {
+        "required": "Please make a selection"
```
Comment (malcolmhire) [inline locales/en/deferral-default-date.json:8]:
```suggestion
        "required": "Select 'yes' if the claimant wants to defer from their State Pension date"
```

**Review by malcolmhire — CHANGES_REQUESTED** [review]

File: app/routes/changes-enquiries/deferral/functions.js (hunk 1 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/deferral/functions.js:112]: done

File: app/routes/changes-enquiries/deferral/functions.js (hunk 2 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/deferral/functions.js:106]: done

File: app/routes/changes-enquiries/award/functions.js (hunk 5 above)
Comment (alittlecross) [reply app/routes/changes-enquiries/award/functions.js:15]: I'm not the best like.

File: app/views/pages/changes-enquiries/deferral/confirm.html [hunk 22]
```diff
@@ -0,0 +1,24 @@
+{% extends 'pages/changes-enquiries/layout.html' %}
+
+{% from "govuk/components/back-link/macro.njk" import govukBackLink %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+
+{% block pageTitle %}{{ t("deferral-confirm:header") }} - Changes and enquiries{% endblock %}
```
[the remark this answers is not in the document]
Comment (alittlecross) [reply app/views/pages/changes-enquiries/deferral/confirm.html:6]: I think I assumed death must have been a recent journey since this kinda ran parallel.

Fixed now.

File: app/views/pages/changes-enquiries/deferral/date-request-received.html [hunk 23]
```diff
@@ -0,0 +1,51 @@
+{% extends 'pages/changes-enquiries/layout.html' %}
+
+{% from "govuk/components/back-link/macro.njk" import govukBackLink %}
+{% from "govuk/components/date-input/macro.njk" import govukDateInput %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+
+{% block pageTitle %}{{ t("deferral-date-request-received:header") }} - Changes and enquiries{% endblock %}
```
[the remark this answers is not in the document]
Comment (alittlecross) [reply app/views/pages/changes-enquiries/deferral/date-request-received.html:7]: done

File: app/views/pages/changes-enquiries/deferral/default-date.html (hunk 9 above)
Comment (alittlecross) [reply app/views/pages/changes-enquiries/deferral/default-date.html:8]: done

File: locales/en/deferral-default-date.json (hunk 21 above)
Comment (alittlecross) [reply locales/en/deferral-default-date.json:8]: done

File: locales/en/deferral-date-request-received.json (hunk 20 above)
Comment (alittlecross) [reply locales/en/deferral-date-request-received.json:10]: done

File: locales/en/deferral-date-request-received.json (hunk 19 above)
Comment (alittlecross) [reply locales/en/deferral-date-request-received.json:9]: done

File: locales/en/stop-state-pension.json (hunk 16 above)
Comment (alittlecross) [reply locales/en/stop-state-pension.json:11]: done

File: app/views/pages/changes-enquiries/stop-state-pension/index.html (hunk 10 above)
Comment (alittlecross) [reply app/views/pages/changes-enquiries/stop-state-pension/index.html:7]: done

File: lib/objects/reviewAwardReasonObject.js (hunk 13 above)
Comment (alittlecross) [reply lib/objects/reviewAwardReasonObject.js:76]: done

File: app/views/pages/changes-enquiries/deferral/date-request-received.html [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+{% block innerContent %}
+  {{ govukBackLink({
+    text: t("app:link.back"),
+    href: backLink,
+    classes: "govuk-!-margin-top-0 govuk-!-margin-bottom-6"
+  }) }}
+
+  <h2 class="govuk-heading-l">{{ t("deferral-date-request-received:h2") }}</h2>
+
+  <form method="post" action={{ formAction }} autocomplete="off">
```
[the remark this answers is not in the document]
Comment (alittlecross) [reply app/views/pages/changes-enquiries/deferral/date-request-received.html:18]: done

File: app/views/pages/customer/add.html (hunk 11 above)
Comment (alittlecross) [reply app/views/pages/customer/add.html:73]: fixed

File: locales/en/add.json (hunk 14 above)
Comment (alittlecross) [reply locales/en/add.json:5]: done

File: locales/en/add.json (hunk 15 above)
Comment (alittlecross) [reply locales/en/add.json:57]: done
